src/analysis/calculus/conformal.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2021 Yourong Zang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yourong Zang
+-/
+import analysis.normed_space.conformal_linear_map
+
+/-!
+# Conformal Maps
+
+A continuous linear map between real normed spaces `X` and `Y` is `conformal_at` some point `x`
+if it is real differentiable at that point and its differential `is_conformal_linear_map`.
+
+## Main results
+
+* `conformal_at`: the main definition
+* The conformality of the composition of two conformal maps, the identity map
+  and multiplications by nonzero constants
+* `conformal_at_iff`: an equivalent definition of the conformality
+* `conformal_at.preserves_angle`: if a map is conformal at `x`, then its differential
+                                  preserves all angles at `x`
+* `conformal_factor_at`: the conformal factor of a conformal map at some point
+
+## Tags
+
+conformal
+
+## Warning
+
+The definition of conformality in this file does NOT require the maps to be orientation-preserving.
+Maps such as the complex conjugate is considered as a conformal map.
+-/
+
+noncomputable theory
+
+variables {E F : Type*} [inner_product_space ℝ E] [inner_product_space ℝ F]
+  {X Y Z : Type*} [normed_group X] [normed_group Y] [normed_group Z]
+  [normed_space ℝ X] [normed_space ℝ Y] [normed_space ℝ Z]
+
+section loc_conformality
+
+open linear_isometry continuous_linear_map
+open_locale real_inner_product_space
+
+/-- A map `f` is said to be conformal if it has a conformal differential `f'`. -/
+def conformal_at (f : X → Y) (x : X) :=
+∃ (f' : X →L[ℝ] Y), has_fderiv_at f f' x ∧ is_conformal_map f'
+
+lemma conformal_at_id (x : X) : conformal_at id x :=
+⟨id ℝ X, has_fderiv_at_id _, 1, one_ne_zero, id, by ext; simp⟩
+
+lemma conformal_at_const_smul {c : ℝ} (h : c ≠ 0) (x : X) :
+  conformal_at (λ (x': X), c • x') x :=
+⟨c • continuous_linear_map.id ℝ X,
+  has_fderiv_at.const_smul (has_fderiv_at_id x) c, is_conformal_map_const_smul h⟩
+
+lemma conformal_at_of_conformal_fderiv {f : X → Y} {x : X}
+  (h : differentiable_at ℝ f x) (H : is_conformal_map (fderiv ℝ f x)) : conformal_at f x :=
+⟨fderiv ℝ f x, h.has_fderiv_at, H⟩
+
+/-- A real differentiable map `f` is conformal at point `x` if and only if
+    its differential `f'` at that point scales any inner product by a positive scalar. -/
+lemma conformal_at_iff {f : E → F} {x : E} {f' : E →L[ℝ] F}
+  (h : has_fderiv_at f f' x) : conformal_at f x ↔ ∃ (c : ℝ) (hc : 0 < c),
+  ∀ (u v : E), ⟪f' u, f' v⟫ = (c : ℝ) * ⟪u, v⟫ :=
+begin
+  split,
+  { rintros ⟨f₁, hf, c₁, hc₁, li, hf₁⟩,
+    exact (hf.unique h) ▸ (is_conformal_map_iff f₁).mp ⟨c₁, hc₁, li, hf₁⟩, },
+  { rintros ⟨c, hc, huv⟩,
+    exact ⟨f', h, (is_conformal_map_iff f').mpr ⟨c, hc, huv⟩⟩, },
```

Comment:
Can you reformulate here so that there's no `split`?  See my comment https://github.com/leanprover-community/mathlib/pull/8367#discussion_r673162149

Reply:
I am really bad at dealing with `iff` lemmas. I particularly don't know how to deal with the different `rcases` for different directions.

Reply:
This is a great place to use your new lemma `conformal_at_iff_is_conformal_map_fderiv`:
```lean
lemma conformal_at_iff' {f : E → F} {x : E} :
  conformal_at f x ↔ ∃ c > 0, ∀ (u v : E), ⟪fderiv ℝ f x u, fderiv ℝ f x v⟫ = c * ⟪u, v⟫ :=
by rw [conformal_at_iff_is_conformal_map_fderiv, is_conformal_map_iff]

lemma conformal_at_iff {f : E → F} {x : E} {f' : E →L[ℝ] F} (h : has_fderiv_at f f' x) :
  conformal_at f x ↔ ∃ c > 0, ∀ (u v : E), ⟪f' u, f' v⟫ = c * ⟪u, v⟫ :=
by simp [conformal_at_iff', h.fderiv]
```
- You should make a choice about which one gets to be named `conformal_at_iff` without the `'`: the `fderiv` version or the `f'` version?  Then use the same choice in naming the two versions of `conformal_factor_at_inner_eq_mul_inner`.
- I have adjusted the wording slightly, from `∃ (c : ℝ) (hc : 0 < c)` to `∃ c > 0`, to be consistent with `is_conformal_map_iff` in the other file.
- The first lemma is basically an `iff` version of `conformal_factor_aux`, so you can then delete that lemma and use `conformal_at_iff'.mp` everywhere instead.

Reply:
Thank you!!

Reply:
It seems that `classical.some` cannot choose properly from a statement like `∃ (c : ℝ) (hc : 0 < c), prop`; it will only obtain a random real number. So I write all `conformal_at`/`is_conformal_map_iff` as `... ↔ ∃ (c : ℝ), 0 < c \and ...`. This reduces one lemma at the cost of the conciseness of four statements.

Reply:
What you have done is fine.  For future reference, to extract the `hc` you do something like
```lean
classical.some $ classical.some_spec $ conformal_at_iff'.mp h
```